web/src/compose_recipient.js

```diff
@@ -244,23 +263,52 @@ function get_options_for_recipient_widget() {
     return options;
 }
 
+function compose_recipient_dropdown_on_show(dropdown) {
+    // Offset to display dropdown above compose.
+    let top_offset = 15;
+    const window_height = window.innerHeight;
+    const search_box_and_padding_height = 50;
+    // pixels above compose box.
+    const recipient_input_top = $("#compose_recipient_selection_dropdown").offset().top;
+    const top_space = recipient_input_top - top_offset - search_box_and_padding_height;
+    // pixels below compose starting from top of compose box.
+    const bottom_space = window_height - recipient_input_top - search_box_and_padding_height;
+    // Show dropdown on top / bottom based on available space.
+    let placement = "top-start";
+    if (bottom_space > top_space) {
+        placement = "bottom-start";
+        top_offset = -30;
+    }
+    const offset = [-10, top_offset];
+    dropdown.setProps({placement, offset});
+    const max_height = Math.min(
+        dropdown_widget.DEFAULT_DROPDOWN_MAX_HEIGHT,
+        Math.max(top_space, bottom_space),
+    );
+    const $popper = $(dropdown.popper);
+    $popper.find(".dropdown-list-wrapper").css("max-height", max_height + "px");
+}
+
+export function open_compose_recipient_dropdown() {
+    $("#compose_select_recipient_widget").trigger("click");
+}
+
 export function initialize() {
-    const opts = {
-        widget_name: "compose_select_recipient",
-        data: get_options_for_recipient_widget(),
-        default_text: $t({defaultMessage: "Select a stream"}),
-        value: null,
-        on_update: on_compose_select_recipient_update,
-    };
-    compose_recipient_widget = new DropdownListWidget(opts);
-    compose_recipient_widget.setup();
-
-    $("#compose_select_recipient_widget").on("select", (e) => {
-        // We often focus on input fields to bring the user to fill it out.
-        // In this situation, a focus on the dropdown div opens the dropdown
-        // menu so that the user can select an option.
-        open_compose_stream_dropup();
-        e.stopPropagation();
+    dropdown_widget.setup(
+        {
+            target: "#compose_select_recipient_widget",
+            on_show_callback: compose_recipient_dropdown_on_show,
+        },
+        get_options_for_recipient_widget,
+        item_click_callback,
+    );
+
+    $("#compose_recipient_selection_dropdown").on("keydown", (e) => {
+        if (e.key === "Enter") {
+            open_compose_recipient_dropdown();
+            e.stopPropagation();
+            e.preventDefault();
+        }
```

Comment:
What does this handler do? I would have thought that if you hit `Esc` to close it and then `Enter` again after, the idea would be that this would make that reopen it, but it appears the `Esc` step doesn't leave focus on this button anyway, so I'm not sure how to focus it without opening it.

Reply:
Fixed.

Reply:
Also, we are trying to transfer the enter press from `compose_recipient_selection_dropdown` to `compose_select_recipient_widget` here.